sycl-fusion/passes/kernel-fusion/SYCLSpecConstMaterializer.h

```diff
@@ -0,0 +1,127 @@
+  size_t SpecConstDataSize;
+  SmallVector<unsigned char> DebugSpecConstData;
+};
+
+///
+/// Pass to materialize specialization constants. Specialization constants
+/// represent constants whose values can be set dynamically during execution of
+/// the SYCL application. The values of these constants are fixed when a SYCL
+/// kernel function is invoked, and they do not change during the execution of
+/// the kernel. This pass receives the values of all specialization constants
+/// used by a kernel and materializes them as concrete types. This is done in
+/// order to be able to enable other optimization opportunities (SCCP, SROA and
+/// CSE), we do not track instructions that can be removed as a result of
+/// materialization, as the pipeline runs DCE pass afterwords.
+class SYCLSpecConstMaterializer
+    : public PassInfoMixin<SYCLSpecConstMaterializer> {
+public:
+  SYCLSpecConstMaterializer() : SpecConstData(nullptr), SpecConstDataSize(0) {}
+
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &);
+
+private:
+  // Main entry point, checks for implicit specialization constant kernel
+  // argument and, if present, performs the optimizations.
+  PreservedAnalyses handleKernel(Function &Kernel);
+
+  bool readMetadata();
+
+  // Collects all the uses of the specialization constant kernel argument.
+  // This results with TypesAndOffsets and LoadsToType being populated.
+  void populateUses(Argument *A);
+
+  // Use TypesAndOffsets to allocate global variables of given types which get
+  // initialized with value taken from the specialization constant blob at a
+  // given offset.
+  void allocateSpecConstant(StringRef KernelName);
+
+  // Re-write uses of loads from the specialization constant kernel argument to
+  // the global variable.
+  void fixupSpecConstantUses();
+
+  // Walk down all uses of a given GEP instruction in order to find loads from
+  // the offsetted pointer.
+  SmallVector<LoadInst *> collectGEPsLoads(GetElementPtrInst *GEP);
+
+  // Helper to report debug message (if enabled) and reset the state.
+  void reportAndReset();
+
+private:
+  // Flag enabling debug output, guarded by: SYCL_MATERIALIZER_DEBUG environment
+  // variable.
+  static const bool IsDebug;
+
+  // Run time known values of specialization constants passed from SYCL rt,
+  // data pointer and size.
+  const unsigned char *SpecConstData;
+  size_t SpecConstDataSize;
+
+  // Module the current function belongs to.
+  Module *Mod{nullptr};
+
+  // Type of the specialization constant and the offset into the SpecConstBlob,
+  // at which the value is located.
+  using TypeAtOffset = std::pair<Type *, uint64_t>;
+  // Helper allowing sorting of TypeAtOffset containers.
+  struct TypeAtOffsetCmp {
+    bool operator()(const TypeAtOffset &LHS, const TypeAtOffset &RHS) const {
+      if (std::get<0>(LHS) == std::get<0>(RHS))
+        return std::get<1>(LHS) < std::get<1>(RHS);
+      return std::get<0>(LHS) < std::get<0>(RHS);
+    }
+  };
```

Comment:
Is this needed? `std::pair` already has a lexicographical comparison operator.

Reply:
You are right, not needed. Removed. Thank you.